rfd/0017-supply-chain-hardening.md

```diff
@@ -0,0 +1,59 @@
+# Supply Chain Hardening
+
+To prevent supply chain attacks take control over all software distribution channels for Teleport and 
+only distribute signed software.
+
+## Why 
+
+Have a cryptographicaly-verifiable audit trail of all software changes and releases 
+and protect all software distribution channels.
+
+## How
+
+### Github
+
+Require [signed commits](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification) 
+for protected branches `master` and `branch/x.y`.
+
+Turn on commit signing for users and bots following steps 
+[here](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification)
+
+**Github Mirror**
+
+Self host git repository for Teleport OSS, Enterprise, Cloud and webapps repositories.
+To allow outside collaboration, allow pull requests, forks, but process all commits using CI/CD pipeline 
```

Comment:
Sorry, I don't understand how this works as described and why you can't use github as a source of truth?

Reply:
I have been worried about their ability to alter content and slow response times. It seems that they have modified the content on the generated tarball and our ticket is waiting for 3+ days for any response. If there was a security incident, that will put is in dangerous territory. 

To mitigate that, I was thinking to keep a read only mirror that is used by folks to fork the content, submit PRs, but it won't be used by Teleport to source any content for builds.

To make this happen, we can have a separate flow for processing pull requests. Folks can still have branches and PRs, but merges on protected branches will be happening outside the github flow by our infrastructure.

May be I'm overthinking this.

Reply:
Well would GPG signing all commits that make it to any release basically prevent githubs ability to modify any of the sources? They basically become a repository, but wouldn't have any ability to change anything without breaking signatures.

For the generated assets, I'm not sure how keeping a duplicate mirror of git somewhere else helps with that problem though. If it's more for backup, it could just be any git client that's keeping a copy instead of having to piece things together from a devs machine.

Reply:
Isn't it the same problem as with log integrity we had? Any actor can remove or rearrange commits and update the branches, unless you sign the entire history?. I'd need to dig deeper to see how github solves this problem with signatures, so maybe I'm wrong.

Reply:
I don't believe it's exactly the same, but I too don't have as good a mental model as needed to be confident in the git properties.

It's possible a signed commit in git also represents the history in some cryptographic secure way, such as chaining the hashes of the commit history. If that's the case, then an adversary hosting the repository has some fairly limited options. They can't change history because that breaks the cryptograghic hashes leading up to the signed commit (if that's a property, I don't know for sure). If they can edit history it might not be a big deal if tools like drone just won't build a commit that isn't signed.

This leaves some fairly limited attack vectors for who's hosting the repo, because the security model is in the data, not the host. So the host could selectively not deliver the last commit to clients, but wouldn't be able to otherwise modify history or files. This would presumably be detectable in a couple of ways, if someone noticed they were missing commits, or there were diverging histories that could no longer become in sync. However it might be a little more subtle, if the release engineer and merge engineer were the same person they may need to notice missing commits in the history. 

Where this gets especially brutal for the host though, is say tools like drone would only work with a signed commit, and if the host like github selectively tried to not return the tag... this would be very detectable as the release engineer wouldn't be able to tell drone to build assets that github as withheld. 

There's lots of ifs and gaps in understanding though and assumptions that we would want to verify. But I believe this is the model signed commits are meant to protect against, as it is the security model developed for distributed development of the linux kernel, and the trust is totally minimized on the host and transfer of the repository itself.

Reply:
https://softwareengineering.stackexchange.com/a/212216 appears to imply that the crypto history git keeps is enough, and as such a signed commit does represent the history. That article also links to some other resources worth exploring.

So from a release engineering perspective, it might just be up to the release engineer to delta the commits before signing a tag for any unexpected changes, and the tools used to build assets are only going to verify the signature on the commit anyways, and verify the hash chain matches (which might preclude something like a shallow clone which is fine for releases).

Reply:
@knisbet You are right. If we sign [every commit](https://mikegerwitz.com/2012/05/a-git-horror-story-repository-integrity-with-signed-commits) and tag (https://stackoverflow.com/questions/5663733/what-is-the-point-in-signing-tags-in-git) we will make sure it won't be possible to alter the tag or rearrange/delete commits associated with a specific tag signed by our CI/CD.

So it seems that we can use github in non-readonly mode as long as we:

* Make sure developers sign every commit with Yubikey
* For external contributors, one of developers would have to amend the commit to sign `git commit -S --amend` as the article suggests
* Drone signs the tag and makes sure that every commit in the tag is signed by one of the valid public keys from the team.  
* Remove the generated `tar.gz` from Github

Reply:
I don't get the part about drone signing the tag. To me that adds surface area with a benefit I must be missing. It seems to imply having a system floating around with signing keys not actually on an HSM like a yubikey / can be told to remotely trust and sign something that external parties will trust.

Also, something else to keep in mind is this methodology makes merges and keeping the history clean a bit more difficult. A squash merge within github can't be used for example, unless maybe someone is going in and amending the mainline commits. So developers have to be careful to keep their branch histories clean, or we'd need some pretty good guidelines on cleaning up before PR.

Reply:
If we don't sign the tag, someone with write access to the repository can delete the tag and create a similar one with the older commit (without security patch). So we'd like to guarantee integrity of the release, the tag have to be signed too.

Reply:
I'm not saying don't sign the tag, why should drone specifically be the signer? Why not the release engineer for example?

Reply:
One idea is that before signing the tag, Drone could verify that signatures of all the commits in the release are from trusted public keys.

Reply:
I would expect no matter who is signing the release tag, that they should be responsible for verifying the contents.

Reply:
That's fair. However, devs could forget to verify the commits or make a mistake. 

I thought we might automate the process to avoid such mistakes. If ask Drone to do verification as a part of release process before creating a signed tag, we can state the invariant: `that tag is only created if all commits since last signed tag are from trusted parties`.

Reply:
Well, that's where having good rock solid processes and tooling come into play. If there is a role within the team for doing release engineering, with the tools required to verify signatures (and even contents) of releases, with checklists and other related disciplines to reduce the risk of someone making a mistake. It might also possible with release engineers to eventually get to an m:n model, with multiple individuals making checks on each other with multiple signatures required (I don't know how this works in git).

From a threat modelling perspective it seems like we're trying to solve a problem statement something along the lines of:
1. Asset release tools/systems should have a way to validate the sources provided to them are from designated individuals.
2. An individual (internal or external) should have a way to copy the repository from public hosts and inspect / validate the tag they received is published by Teleport.
3. Both of the above to be delivered by not trusting the repository host.

By automating this through drone, to me implies a release process that looks something like:
1. A release engineer tells drone we need a new release on a specific branch.
   2. Note: I'm still a bit unclear on this, the release engineer tells drone to tag a specific commit by hash? If the specification is by branch, a release engineer will need to later validate that drone did tag the latest commit.
2. Drone checks out the code from the public host, verifies signatures of individual commits, and publishes the signed tag back to the repository host(s).
3. (Optional) Based on above, the release engineer needs to validate the repository host didn't drop the commit when provided to drone.
4. Drone still needs to build the assets, so based on the new tag, drone checks out the code, verifies the signature against it's own key, and proceeds to run normal build and asset publishing to various locations we make the release assets available.


To me having drone sign that tags implies a number of implications:
1. Drone will need additional key handling, with all the normal key handling constraints, in order to tag and publish back to the public repo.
    1. How are those keys stored, backed up, etc.
    2. Who has access to those keys, what happens when an employee leaves, can the keys even be revoked and rotated.
3. Drone itself becomes an attack vector, is there a way to introduce additional configuration to trick drone into using the signing secret in some way it's not supposed to (release it outright, signing an asset or tag it shouldn't). This might be the situation of other secrets as part of the release process, but for this spec I think it's still a consideration from a surface area perspective.
    1. A breach of drones secrets also creates a property, that a particular validator could be the target by selectively changed secrets, especially if we consider a host like github not to be secure. So the signed tag x could be  delivered to drone for build verification while all other users receive an alternate tag. I suspect this would be fairly difficult to detect based on what's described.
4. Drone needs to be kept in sync with all active developers for the company
    1. How should drone react to a developer who has left, has merged commits, that haven't been tagged yet?
5. How do we communicate and configure the public keys of the developers securely?
6. How do we communicate externally the public keys for an individual who wants to checkout and verify a tag?
7. I suspect all developers must be handling their GPG keys securely, as drone will trust a merged commit by any developer???
8. Who has access to drone and the relevant secrets? etc.